lib/floki/raw_html.ex

```diff
@@ -99,13 +110,19 @@ defmodule Floki.RawHTML do
   defp tag_with_attrs(type, attrs, children, padding),
     do: [leftpad(padding), "<", type, ?\s, tag_attrs(attrs) | close_open_tag(type, children)]
 
-  defp close_open_tag(type, []) when type in @self_closing_tags, do: "/>"
-  defp close_open_tag(_type, _), do: ">"
-
-  defp close_end_tag(type, [], _padding) when type in @self_closing_tags, do: ""
+  defp close_open_tag(type, children) do
+    case {type in self_closing_tags(), children} do
+      {true, []} -> "/>"
+      _ -> ">"
+    end
+  end
 
-  defp close_end_tag(type, _, padding),
-    do: [leftpad(padding), "</", type, ">", line_ending(padding)]
+  defp close_end_tag(type, children, padding) do
+    case {type in self_closing_tags(), children} do
+      {true, []} -> ""
```

Comment:
It's slightly cheaper to have an empty list than an empty string. Since we are using IO data here, this is possible:

```suggestion
      {true, []} -> []
```